File: cross-platform/src/core/level.ts
/**
 * Focus levels and the signals they are driven at, ported from
 * `Level.swift` and `SignalProfile.swift`.
 *
 * A level names a destination and carries its default noise bed; a *signal
 * profile* is what was actually measured off a tape. Where a level names a
 * profile, the measurement wins over the configured pair — which is the whole
 * reason the two are separate types rather than one.
 */

export interface LevelBed { pink: number; white: number }

export interface Level {
  key: string;
  name: string;
  beatHz: number;
  carrier: number;
  signalProfile?: string;
  bed: LevelBed;
  layers: number[];
  rampSeconds: number;
  /** False when `beatHz` is a placeholder rather than a value carried over
   *  from a tuned level. Nothing may render at an unverified beat without
   *  saying so. */
  beatVerified: boolean;
  /** Present, and non-empty, when this level is populated by minds the
   *  listener did not choose — what the affirmation's protective clause is
   *  actually gating. */
  exposure?: string;
  /** The listener's own working description. Theirs to correct. Defaults to
   *  "" like every other string field here, never absent. */
  notes: string;
  /** What the Monroe Institute publishes about this level — a baseline to be
   *  disproven, never the last word. Kept apart from `notes` so neither can
   *  quietly overwrite the other. */
  published: string;
}

/** True when this level is populated by minds the listener did not choose. */
export const isExposure = (level: Level): boolean =>
  (level.exposure ?? "") !== "";

export interface SignalHold {
  start: number;
  end: number;
  carrier: number;
  /** The right ear is `carrier + beat`. */
  beat: number;
  gain: number;
  confidence: number;
}

export const holdDuration = (h: SignalHold): number => Math.max(0, h.end - h.start);

export interface SignalProfile {
  id: string;
  level?: string;
  duration: number;
  holds: SignalHold[];
}

/**
 * The stable pair which carries most of a tape: duration weighted by measured
 * gain, then confidence.
 *
 * Deliberately ignores brief loud detections — the live bed needs a dependable
 * foundation, not every FFT transient from the source recording.
 */
export function dominantHold(p: SignalProfile): SignalHold | undefined {
  let best: SignalHold | undefined;
  for (const h of p.holds) {
    if (!best) { best = h; continue; }
    const a = holdDuration(best) * best.gain;
    const b = holdDuration(h) * h.gain;
    // `max(by:)` keeps the later element only when it is strictly greater,
    // which is what makes this stable against reordering equal holds.
    if (a === b ? best.confidence < h.confidence : a < b) best = h;
  }
  return best;
}

export interface LevelSignal { carrier: number; beat: number; source?: string }

/** A measured profile wins over the level's configured pair. */
export function resolvedSignal(level: Level, profiles: SignalProfile[]): LevelSignal {
  if (level.signalProfile) {
    const profile = profiles.find(p => p.id === level.signalProfile);
    const hold = profile ? dominantHold(profile) : undefined;
    if (hold) return { carrier: hold.carrier, beat: hold.beat, source: profile!.id };
  }
  return { carrier: level.carrier, beat: level.beatHz };
}

/** What a level with no configuration at all is driven at. */
export const fallbackSignal: LevelSignal = { carrier: 100, beat: 0 };

File: Sources/GatewayCore/Level.swift
import Foundation

/// A Focus level. Configuration, never code -- nothing switches on a level key.
public struct Level: Codable, Identifiable, Hashable, Sendable {
    public var key: String
    public var name: String
    public var beatHz: Double
    public var carrier: Double
    /// The measured tape profile whose sustained primary pair should drive this
    /// level. Nil keeps the authored `carrier` / `beatHz` fallback. The choice
    /// lives in levels.json so changing evidence never requires an engine
    /// switch on a Focus key.
    public var signalProfile: String?
    public var bed: Bed
    public var layers: [Double]
    /// Seconds to ramp into this level's beat. A step change is audible; ramps are not.
    public var rampSeconds: Double
    /// The user's own working description. Theirs to correct.
    public var notes: String
    /// What the Monroe Institute publishes about this level. A baseline to be
    /// disproven, never the last word: F26 is published as a Belief System
    /// Territory and is experienced as the dark void, and both readings have to
    /// stay visible for the disagreement to be useful. Kept apart from `notes`
    /// so neither can quietly overwrite the other.
    public var published: String
    /// Why arriving here exposes the listener to minds that are not their own
    /// and not well — nil for every level where it does not.
    ///
    /// This is not a difficulty rating and not a warning label. It answers one
    /// question, and the protective clause of the 1977 Affirmation is the only
    /// thing that reads it: *does going here put you in front of something you
    /// did not choose?* F10 through F21 are states of your own mind, however
    /// deep; nobody else is there. F22 through F26 and the Gathering are
    /// populated, and the population did not agree to meet you.
    ///
    /// Kept as prose rather than a `Bool` because the reason is the useful
    /// part: it is what a session can show the listener, and it is what a
    /// future editor needs in order to disagree with the marking.
    public var exposure: String?

    /// False when `beatHz` is a placeholder rather than a value carried over from
    /// a tuned level. Nothing may render at an unverified beat without saying so.
    public var beatVerified: Bool

    public var id: String { key }

    public struct Bed: Codable, Hashable, Sendable {
        public var pink: Double
        public var white: Double
        public init(pink: Double, white: Double) { self.pink = pink; self.white = white }
    }

    public init(key: String, name: String, beatHz: Double, carrier: Double = 110,
                signalProfile: String? = nil,
                bed: Bed = .init(pink: 0.28, white: 0.08), layers: [Double] = [],
                rampSeconds: Double = 20, notes: String = "",
                published: String = "", exposure: String? = nil,
                beatVerified: Bool = true) {
        self.key = key; self.name = name; self.beatHz = beatHz; self.carrier = carrier
        self.signalProfile = signalProfile
        self.bed = bed; self.layers = layers; self.rampSeconds = rampSeconds; self.notes = notes
        self.published = published; self.exposure = exposure
        self.beatVerified = beatVerified
    }

    /// Hand-written and hand-editable, so a missing key must not take the whole
    /// library down with it. Swift's synthesised decoding ignores property
    /// defaults and throws instead -- and `levels.json` is a file the user is
    /// expected to open and edit.
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        key = try c.decode(String.self, forKey: .key)
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? key
        beatHz = try c.decodeIfPresent(Double.self, forKey: .beatHz) ?? 0
        carrier = try c.decodeIfPresent(Double.self, forKey: .carrier) ?? 110
        signalProfile = try c.decodeIfPresent(String.self, forKey: .signalProfile)
        bed = try c.decodeIfPresent(Bed.self, forKey: .bed) ?? .init(pink: 0.28, white: 0.08)
        layers = try c.decodeIfPresent([Double].self, forKey: .layers) ?? []
        rampSeconds = try c.decodeIfPresent(Double.self, forKey: .rampSeconds) ?? 20
        notes = try c.decodeIfPresent(String.self, forKey: .notes) ?? ""
        published = try c.decodeIfPresent(String.self, forKey: .published) ?? ""
        exposure = try c.decodeIfPresent(String.self, forKey: .exposure)
        beatVerified = try c.decodeIfPresent(Bool.self, forKey: .beatVerified) ?? true
    }

    public var allBeats: [Double] { [beatHz] + layers }

    /// True when this level is populated by minds the listener did not choose.
    public var isExposure: Bool { !(exposure ?? "").isEmpty }
}
